### PR TITLE
[WIP] Share the TF-IDF doc ranker

### PR DESCRIPTION
**Pull request by EricMichaelSmith**
**Patch description**
Share TfidfDocRanker. I made this in order to avoid reloading the data at `tfidf_path` for each copy of the TfidfDocRanker in my batch. This is used when adding knowledge to the Transformer polyencoder in https://github.com/fairinternal/ParlAI-Internal/pull/942

**Testing steps**
(NOTE: will retest after simplifying PR in response to comments)
Runs successfully (i.e. losses go down, performance looks reasonable so far) on my devfair with https://github.com/fairinternal/empathetic-dialog/blob/ems-dev/all_in_one_dialogue/_scripts/s2019_09_20__wow_with_knowledge/00_local_test.sh and on SLURM with https://github.com/fairinternal/empathetic-dialog/blob/ems-dev/all_in_one_dialogue/_scripts/s2019_09_20__wow_with_knowledge/01_launch_sweep.py

### CODE REVIEW

File: parlai/agents/tfidf_retriever/tfidf_retriever.py [hunk 1]
```diff
@@ -163,9 +162,15 @@ def __init__(self, opt, shared=None):
...
-            )
+            if shared is None:
+                doc_ranker_opt = {
+                    'model': 'tfidf_retriever/tfidf_doc_ranker',
+                    'tfidf_path': opt['retriever_tfidfpath'],
+                    'strict': False,
+                }
+                self.ranker = create_agent(doc_ranker_opt)
+            else:
+                self.ranker = create_agent_from_shared(shared['doc_ranker_opt'])
```
Comment (emilydinan) [inline parlai/agents/tfidf_retriever/tfidf_retriever.py]: probably best to just share the ranker here

File: parlai/agents/tfidf_retriever/tfidf_doc_ranker.py [hunk 2]
```diff
@@ -17,32 +17,52 @@
 
 from . import utils
 from . import tokenizers
+from parlai.core.agents import Agent
 
 logger = logging.getLogger(__name__)
 
 
-class TfidfDocRanker(object):
+class TfidfDocRankerAgent(Agent):
```
Comment (stephenroller) [inline parlai/agents/tfidf_retriever/tfidf_doc_ranker.py]: An agent has to implement observe/act. Probably you just want to share the object directly in the user.

**Comment by EricMichaelSmith** [comment]
Thanks @stephenroller and @emilydinan ! Okay, makes sense - I'll make these changes soon (hopefully sometime this week)

**Comment by EricMichaelSmith** [comment]
Thanks @emilydinan and @stephenroller for the feedback! I've made these changes - how does it look now?

**Pull request merged**